src/ray/core_worker/experimental_mutable_object_manager.h

```diff
@@ -206,20 +191,18 @@ class MutableObjectManager {
   FRIEND_TEST(MutableObjectTest, TestReadAcquireDuringFailure);
   FRIEND_TEST(MutableObjectTest, TestReadMultipleAcquireDuringFailure);
 
-  // TODO(swang): Access to these maps is not threadsafe. This is fine in the
-  // case that all channels are initialized before any reads and writes are
-  // issued, but may not work if channels are added/removed during run time
-  // (i.e., if there are multiple DAGs).
-  // TODO(jhumphri): If we do need to synchronize accesses to these maps, we may want to
+  // TODO(jhumphri): If we do need to synchronize accesses to this map, we may want to
   // consider using RCU to avoid synchronization overhead in the common case.
-  // These two maps hold the channels for readers and writers of mutable objects.
-  absl::flat_hash_map<ObjectID, ReaderChannel> reader_channels_;
-  absl::flat_hash_map<ObjectID, WriterChannel> writer_channels_;
+  // This map holds the channels for readers and writers of mutable objects.
+  absl::Mutex channel_lock_;
+  absl::flat_hash_map<ObjectID, Channel> channels_;
 
   // This maps holds the semaphores for each mutable object. The semaphores are used to
   // (1) synchronize accesses to the object header and (2) synchronize readers and writers
   // of the mutable object.
   absl::flat_hash_map<ObjectID, PlasmaObjectHeader::Semaphores> semaphores_;
+
+  absl::Mutex destructor_lock_;
```

Comment:
Can you add a comment explaining why we need these locks? Who is are the calling threads, etc.

Also, should we just use one lock? Seems less error-prone. It'd also be good to annotate the class members like `channels_` with `ABSL_GUARDED_BY`.

Reply:
Could please you elaborate on how to do this with one lock? The two locks seem to serve different purposes. Thanks.

Also, do you want me to annotate all class members with ABSL_GUARDED_BY for the destructor lock? The lock is basically meant to protect all internal state of the class.